### PR TITLE
Convert remaining code to pybind11

### PR DESCRIPTION
**Pull request by QuLogic**
## PR summary

Now that we are building with pybind11 in all extensions, we can drop/convert several items that were used across extensions. Namely, everything in `py_adaptors` (which was shared between `_path` and `_backend_agg`) can now get a type caster, all the remaining `array_view` can be moved to `py::array_t`, and the `ClipPath` converter issue from #27011 has been fixed.

Additionally, I moved the pybind11 type casters from `py_converters_11.h` to the files that define the related types. This keeps things a bit more consolidated, and avoids the weird `#ifdef` macro magic that was needed before.

## PR checklist

- [n/a] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [n/a] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Review by ksunden — APPROVED** [review]
I would like to get this in for 3.10, so if we could get a second review, that would be great

File: src/_backend_agg_basic_types.h
Comment (anntzer) [inline src/_backend_agg_basic_types.h:155]: I don't think you need the default here (either the next line will throw or it won't)?

File: src/_backend_agg_basic_types.h
Comment (anntzer) [inline src/_backend_agg_basic_types.h:170]: Would probably write as something like (untested...)
```c++
auto [path, trans] = src.cast<std::pair<std::optional<mpl::PathIterator>, agg::trans_affine>>();
if (path) { value.path = *path; }
value.trans = trans;
```
(note that the current implementation doesn't actually check the length of the tuple).
Ditto for dashes below.

File: src/_backend_agg_basic_types.h
Comment (QuLogic) [reply src/_backend_agg_basic_types.h:155]: There wasn't a default in `line_cap_e`, so this must have been some leftover that wasn't cleaned up.

File: src/_backend_agg_basic_types.h
Comment (QuLogic) [reply src/_backend_agg_basic_types.h:170]: I thought I had tried this and ran into something weird with `std::optional`, but it seems to work. Maybe it had something to do with parameters directly.

**Comment by tacaswell** [comment]
This looks like some of the grayscale have picked up a off-by-one rounding issue.  Seeing this on other PRs as well.

**Review by greglucas — APPROVED** [review]
I am not an expert on the pybind11 changes, but I did peruse this and it all looks reasonable to a non-expert. I reran the failing 313t tests and those now pass. I believe the other failures in Azure are unrelated and failing on other tests too.

I believe that it will be easier to review incremental updates after this is in if there is anything else to follow up on.

Great work on this @QuLogic! This was no small feat.

**Pull request merged**
